File: backend/generateURL.js

// NOTES: 
// Add { "type": "module" } to package.json and change the name of the imported file 
// To run, execute "node generateURL" in a terminal
// String-type query parameters are case insensitive

import file from './Example5.json' assert { type : 'json' };
let URL = "https://krishnam.tlcengine.com/search/?orderby=newest&orderbydirection=DESC&page_number=1&searchmap=true&skip=0&view=maplist";

function JSON_to_URL() {

    for (let key in file) {

        try {

            switch (key) {

                case "acres":
                    if (file[key][0] != null) URL += "&acresmin=" + file[key][0];
                    if (file[key][1] != null) URL += "&acresmax=" + file[key][1];
                    break;
            
                case "city":
                    if (file[key] != null) URL += "&city=" + encodeURIComponent(file[key].trim());
                    break;

                case "homeFeatures":
                    for (let i in file[key]) URL += "&homeFeatures=" + encodeURIComponent(file[key][i].trim());
                    break;

                case "listingprice":
                    if (file[key][0] != null) URL += "&listingpricemin=" + file[key][0];
                    if (file[key][1] != null) URL += "&listingpricemax=" + file[key][1];
                    break;

                case "livingarea":
                    if (file[key][0] != null) URL += "&livingareamin=" + file[key][0];
                    if (file[key][1] != null) URL += "&livingareamax=" + file[key][1];
                    break;

                case "lotfeatures":
                    for (let i in file[key]) URL += "&lotfeatures=" + encodeURIComponent(file[key][i].trim());
                    break;

                case "propertystyle":
                    for (let i in file[key]) URL += "&propertystyle=" + encodeURIComponent(file[key][i].trim());
                    break;

                case "propertytype":
                    for (let i in file[key]) URL += "&propertytype=" + encodeURIComponent(file[key][i].trim());
                    break;

                case "stories":
                    if (file[key] != null) URL += "&stories=" + encodeURIComponent(file[key].trim());
                    break;

                case "streetaddress":
                    if (file[key] != null) URL += "&streetaddress=" + encodeURIComponent(file[key].trim());
                    break;

                case "yearbuilt":
                    if (file[key][0] != null) URL += "&yearbuiltmin=" + file[key][0];
                    if (file[key][1] != null) URL += "&yearbuiltmax=" + file[key][1];
                    break;

                default:
                    if (typeof(file[key]) == "string") file[key] = file[key].trim(); 
                    if (file[key] != null) URL += "&" + key + "=" + file[key];
                    break;
            }

        } catch (error) {
            console.error(key + " --> " + error);
        }
    }

    return URL;
}

console.log(JSON_to_URL());

File: backend/generateURL.py
import json
import urllib.parse

# Base URL
# Shold put this in the .env
url = "https://krishnam.tlcengine.com/search/?orderby=newest&orderbydirection=DESC&page_number=1&searchmap=true&skip=0&view=maplist"


def json_to_url(data):
    global url
    for key, value in data.items():
        try:
            if key == "acres":
                if value[0] is not None:
                    url += f"&acresmin={value[0]}"
                if value[1] is not None:
                    url += f"&acresmax={value[1]}"
            elif key == "city":
                if value:
                    url += f"&city={urllib.parse.quote(value.strip())}"
            elif key == "homeFeatures":
                for feature in value:
                    url += f"&homeFeatures={urllib.parse.quote(feature.strip())}"
            elif key == "listingprice":
                if value[0] is not None:
                    url += f"&listingpricemin={value[0]}"
                if value[1] is not None:
                    url += f"&listingpricemax={value[1]}"
            elif key == "livingarea":
                if value[0] is not None:
                    url += f"&livingareamin={value[0]}"
                if value[1] is not None:
                    url += f"&livingareamax={value[1]}"
            elif key == "lotfeatures":
                for feature in value:
                    url += f"&lotfeatures={urllib.parse.quote(feature.strip())}"
            elif key == "propertystyle":
                for style in value:
                    url += f"&propertystyle={urllib.parse.quote(style.strip())}"
            elif key == "propertytype":
                for ptype in value:
                    url += f"&propertytype={urllib.parse.quote(ptype.strip())}"
            elif key == "stories":
                if value:
                    url += f"&stories={urllib.parse.quote(value.strip())}"
            elif key == "streetaddress":
                if value:
                    url += f"&streetaddress={urllib.parse.quote(value.strip())}"
            elif key == "yearbuilt":
                if value[0] is not None:
                    url += f"&yearbuiltmin={value[0]}"
                if value[1] is not None:
                    url += f"&yearbuiltmax={value[1]}"
            else:
                if isinstance(value, str):
                    value = value.strip()
                if value is not None:
                    url += f"&{key}={value}"
        except Exception as error:
            print(f"{key} --> {error}")
    return url

# Test Code
# with open("test.json", "r") as file:
#    data = json.load(file)
#final_url = json_to_url(data)
# print(final_url)
